### DISCUSSION TITLE
Defining a sampler with a specific transition rule

### DISCUSSION THREAD

**Discussion by waleed-sh** [netket/netket#1759] [Q&A] [answered]
Hi! My questions is regarding defining a sampler with a custom transition rule. I saw the code for the TwoLocal rule [here](https://netket.readthedocs.io/en/latest/docs/sampler.html#defining-custom-transition-rules-for-metropolis-sampler), but I am still a little lost...

The idea is the following:
I am trying to define a local sampler which somehow only generates states which have a certain relation between different basis elements. For example, say I have a 5 site chain, so basis elements are [a, b, c, d, e]. I want the sampler to somehow only generate states in which, for example, c = -a, and/or b = - d - e. 

In essence, I have a system with a very large Hilbert space, but I know what type of states I should be looking for so I was wondering if I can already narrow the space to be explored at the sampling level. Any help or guidance would be appreciated!

**Comment by PhilipVinc** [comment]
Can you formalise the rule you'd want mathematically? What is your local basis? A spin-1?
It's hard to parse what you mean exactly.

**Reply by waleed-sh** [reply]
My local basis is somewhat like angular momentum, so a $|j>$ basis where the integers $j \in J = [-j_{max} , \dots , j_{max}]$ label the basis, here $j_{max}$ is just an integer. So at every vertex there would be $2j_{max} + 1$ possible local degrees of freedom which are the numbers $J$. 

My systems are defined on oriented graphs. The basis states which happen to contribute to the wave function are ones which have certain characteristics depending on the orientation. This changes from graph to graph, but the general picture can be understood that if a vertex A is connected to a vertex B via an outgoing edge (so $A \rightarrow B$), then the local state on B needs to be equal to but the opposite sign of that on A. This becomes more complicated for the case where several edges are attached to a certain vertex. But this case is for now put aside. 


I'm fine with a hard coded solution to just see how to create a sampler with a transition rule that only samples states where the labels of two user specified basis states have a combined value which is equal to but opposite in sign to a third user specified state (e.g. for a basis $|j_1> \otimes |j_2> \otimes |j_3> \otimes |j_4> \otimes |j_5>$ then the sampler should only search for states where $j_1 + j_2 = - j_4$). I think from there onwards I can generalise it to my specific cases. 

I apologise if this is very ad hoc, but I am trying to avoid a rather lengthy mathematical discussion. Any guidance or examples would be appreciated!

**Reply by PhilipVinc** [reply]
Are you able of writing a function `f`that given a single basis state  and the three indices `I,j,k` gives you all possible states that satisfy that condition?

**Reply by waleed-sh** [reply]
Yes I think this is something possible to do so it shouldn't be a problem.

**Reply by waleed-sh** [reply]
I was thinking about this again and maybe this issue could be solved by defining a constrained Hilbert space? So from the start only allow for states which obey a certain criteria? But from my understanding this can cause issues with samplers which do not respect this criteria so I am not sure if then I will have to either way construct a sampler which does the same thing. Or am I wrong?

**Reply by PhilipVinc** [reply, marked as answer]
What you want is equivalent to what [HamiltonianRule](https://github.com/netket/netket/blob/926aca630546d77b711cd5a1ad507828d42e167b/netket/sampler/rules/hamiltonian.py#L65) does. 

If you look at its source, it [queries the operator for its connected entries](https://github.com/netket/netket/blob/926aca630546d77b711cd5a1ad507828d42e167b/netket/sampler/rules/hamiltonian.py#L101) (this could be replaced by your function computing the connected entries given one), then it picks one, then it [checks the connected components of the proposed site to compute the Hastings log correction, as your proposal might not be simmetrical](https://github.com/netket/netket/blob/926aca630546d77b711cd5a1ad507828d42e167b/netket/sampler/rules/hamiltonian.py#L107).

You'd want to replace calls to `get_conn_flattened` to calls to your own function that lists the connected components given a certain configuration.

--

> I was thinking about this again and maybe this issue could be solved by defining a constrained Hilbert space

Unlikely. Hilbert spaces are used to generate random configurations (like when you initialise your Markov chains) or for full summation things. Samplers might drop out of the constraint if, indeed, they do not respect the constraint. 
The solution is, as above, to write a transition rule that respects it.

**Reply by waleed-sh** [reply]
Thank you for your reply! Just a quick possibly stupid question, what would the return type and the arguments of the `get_conn_flattened`? Because it uses the `get_conn_flattened_closure()` from the operator, which eventually leads to the complicated `get_conn_flattened_kernel()` (if I understand correctly). I'm just trying to see what datatype should the function expect and what is the simplest thing it should return for it to function. 

Thank you for your reply though! This has been tremendously helpful. I think I will go with both a constrained Hilbert space and a sampler respecting this constraint as I also face some issues due to the dimensionality of the space.

**Reply by PhilipVinc** [reply]
Don't look at that stuff, it's overly complicated to work around limitations of numba while maintaining high performance.

Look at this version https://github.com/netket/netket/blob/a1286d79e1e85d0ab113dae9f2610ddcc10fcadb/netket/sampler/rules/hamiltonian.py#L89 
it's equivalent. Just a tad bit slower, but much easier to understand. 

(t's taken from this PR https://github.com/netket/netket/pull/1747 )

**Reply by waleed-sh** [reply]
Great, thanks again! I will give it a try I think it should be straightforward enough.
